### PR TITLE
LNURL-Pay.

### PR DESCRIPTION
**Pull request by nochiel**
This PR depends on [PR 493](https://github.com/breez/breezmobile/pull/518).

--- 

### Todo: 

- [ ] Write tests for `lnurl.go`. 
    - What is the correct way to do this so I can automate it?
- [x] Ensure errors are propagated to `breezmobile` correctly.
    - Is it enough for functions in `api.go` to simply return the error?
- [x] Ensure notifications (`SuccessAction`) are presented to the user correctly.
    - What is the correct way to push a toast/popup to `breezmobile`?
    - [x] Handle `SuccessAction` "aes" tag.
- [ ] Verifications required by spec:
    - [ ]  signatures for each `ChannelUpdate`.
      - How is `ChannelUpdate` signature verification done?
    - [x] use description hash  to verify metadata.
    - [ ] check which parameters can be null before trying to use them.
- [x] Show the user the fee for the payment route.
    - [x] Verify that this works the way standard payments work.
- [x] Upgrade fiatjaf/go-lnurl so that we can use the new rawurl return value from lnurl.HandleLNUrl to get the domain of the host serving the lnurl and display it to the user.
-  [x] Code cleanup.
   - [x] Remove unnecessary comments.

### CODE REVIEW

**Comment by roeierez** [comment]
> Please tell me if I'm on the right track. I'm still getting familiar with the codebase and lightning

Initial pass and the building blocks looks in place. Thank you for this good start.
Regarding the second phase (paying the invoice), I think it is a good thing to return the invoice and payment route to breez mobile and not actually pay from the FinishLNURLPay function.
The reason is since we have all the logic of paying an invoice (including UI, error handling and edge cases) in breez mobile and we can merge into this flow with a given invoice (the class that handles the payment is https://github.com/breez/breezmobile/blob/d39dab1157a564c1567f48afdf7f1f84b0a15db5/lib/widgets/payment_request_dialog.dart). By initiating this dialog with the right parameters (including SendAction with payment request) it will handle all aspects of the payments.
This also let us show the user the fee before initiating the payment.
Let me know if it makes sence or if something goes wrong. Thanks again for this great start.

**Comment by roeierez** [comment]
> Handle `SuccessAction` "aes" tag.
> * Decryption of ciphertext needs payment preimage.
> * The preimage is revealed when the transaction goes on-chain?
> * Are there other conditions when the preimage is revealed so we can use it?

preimage is revealed in lightning after payment is succesfull. In my understanding this let the user decrypt the success action (in case of aes action).
So after payment is succesfull we can decrypt and show the right message but I suggest to handle this after the payment is implemented as we may need to persist this value.

**Comment by roeierez** [comment]
> Verify signatures for each `ChannelUpdate`.
> * How is `ChannelUpdate` signature verification done?

Not sure about the verification process. LND has these channel updates in his graph database internally. Maybe just build a route based on the given nodes and channels and obtain the fees from LND? need to think about it.

**Comment by kingonly** [comment]
@nochiel thanks for this PR! I hope we could make progress with it and to support LNURL-Pay in Breez. Just to clarify: from a bounty standpoint, we do expect a functional solution and not a WIP PR. I hope that's ok.

**Comment by nochiel** [comment]
> @nochiel thanks for this PR! I hope we could make progress with it and to support LNURL-Pay in Breez. Just to clarify: from a bounty standpoint, we do expect a functional solution and not a WIP PR. I hope that's ok.

Yes, certainly. That's why I requested some early feedback so that I don't go too far down the wrong track. Also, there are ux design decisions that I don't know how to answer. A functional solution is the goal for me whether or not there's a bounty. The Breez codebase is attractive to me because it's easy to read and understand and I was looking for a way to learn Lightning⚡ by contributing code.

**Comment by nochiel** [comment]
> Let me know if it makes sense or if something goes wrong. Thanks again for this great start.

Thank you so much for taking the time to give me feedback!

File: account/lnurl.go [hunk 1]
```diff
@@ -200,3 +222,217 @@ func (a *Service) getLNURLAuthKey() (*bip32.Key, error) {
...
+	query.Add("amount", fmt.Sprintf("%d", params.Amount))
+
+	// source: backup/crypto.go
+	nonce := make([]byte, 12)
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		return nil, err
+	}
+	query.Add("nonce", string(nonce))
+
+	/* TODO FINDOUT Should we populate the FromNodes optional parameter?
```
Comment (roeierez) [inline account/lnurl.go:254]: I think it is OK to leave empty for now.

File: account/lnurl.go [hunk 2]
```diff
@@ -200,3 +222,217 @@ func (a *Service) getLNURLAuthKey() (*bip32.Key, error) {
...
+	// a.lnurlSuccessAction = sa // FIXME Handle this in client.
+	return &data.LNUrlPayInfo{
+		Invoice: payResponse2.PR,
+		SuccessAction: _sa,
+	},
+	nil
+
+}
+
+func (a *Service) decryptLNUrlPayMessage(preImage []byte) (m string, err error) {
```
Comment (roeierez) [inline account/lnurl.go:431]: Is this in use?

**Review by roeierez — CHANGES_REQUESTED** [review]
Looks good! 
Some small comments bellow.

File: account/lnurl.go (hunk 2 above)
Comment (nochiel) [reply account/lnurl.go]: > Is this in use?

Yes. This has all the data + invoice that should be displayed to the user before payment confirmation.

File: account/lnurl.go [hunk 3]
```diff
@@ -78,6 +82,24 @@ func (a *Service) HandleLNURL(rawString string) (*data.LNUrlResponse, error) {
 				},
 			},
 		}, nil
+	case lnurl.LNURLPayResponse1:
+		a.log.Info("lnurl.LNURLPayResponse1.")
+		host := ""
+		if u, err := url.Parse(rawurl); err != nil {
+			host = u.Host
```
[the remark this answers is not in the document]
Comment (nochiel) [reply account/lnurl.go]: > Did you mean to return error here?

Fixed. The correct condition is `err == nil`.

File: account/lnurl.go (hunk 2 above)
Comment (nochiel) [reply account/lnurl.go]: > Is this in use?

Not yet. It will be used for SuccessAction.

File: account/lnurl.go [hunk 4]
```diff
@@ -78,6 +82,27 @@ func (a *Service) HandleLNURL(rawString string) (*data.LNUrlResponse, error) {
 				},
 			},
 		}, nil
+	case lnurl.LNURLPayResponse1:
+		a.log.Info("lnurl.LNURLPayResponse1.")
+		host := ""
+		if url, err := url.Parse(rawurl); err == nil {
```
Comment (roeierez) [inline account/lnurl.go:88]:
```suggestion
		if url, err := url.Parse(rawurl); err != nil { 
		  return nil, err
		}
		host := url.Host
```

**Review by roeierez — APPROVED** [review]
Looks good!
Other than one comment bellow I approved.

**Review by roeierez — APPROVED** [review]
LGTM, great work.

**Pull request merged**